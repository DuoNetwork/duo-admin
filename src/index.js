import pf from "./priceSender/priceFeed";
import listenToAcceptPrice from "./listenToAcceptPrice";
import createAccount from "./accounts/createAccounts";
import bitfinexTradeFeedUtil from "./dataFetcher/bitfinex/bitfinexUtil";
import geminiTradeFeedUtil from "./dataFetcher/gemini/geminiUtil";
import krankenTradeFeedUtil from "./dataFetcher/kraken/krakenUtil";
import coinbaseGDAXTradeFeedUtil from "./dataFetcher/gdax/gdaxUtil";

let tool = process.argv[2];

switch (tool) {
	case "pf":
		console.log("starting commitPrice process");
		pf.startFeeding();
		break;
	case "acceptPrice":
		console.log("starting listening to acceptPrice event");
		listenToAcceptPrice.startListening();
		break;
	case "createAccount":
		console.log("starting create accounts");
		createAccount.createAccount(process.argv[3]);
		break;
	case "bitfinex":
		console.log("starting fetchTrade of bitfinex");
		bitfinexTradeFeedUtil.fetchETHTradesByOwnWebSocket();
		break;
	case "gemini":
		console.log("starting fetchTrade of gemini");
		geminiTradeFeedUtil.fetchETHTradesByOwnWebSocket();
		break;
	case "kraken":
		console.log("starting fetchTrade of kraken");
		krankenTradeFeedUtil.startFetching();
		break;

	case "gdax":
		console.log("starting fetchTrade of gdax");
		coinbaseGDAXTradeFeedUtil.startFetching();
		break;
}
