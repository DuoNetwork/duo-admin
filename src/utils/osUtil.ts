import * as os from 'os';

class OsUtil {
	public getHostName() {
		return os.hostname();
	}

	public isWindows() {
		return process.platform === 'win32';
	}
}

const osUtil = new OsUtil();
export default osUtil;
