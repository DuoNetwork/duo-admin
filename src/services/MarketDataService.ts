import child_process from 'child_process';
import apis from '../apis';
import { IOption, ISubProcess } from '../common/types';
import dbUtil from '../utils/dbUtil';
import osUtil from '../utils/osUtil';
import util from '../utils/util';
import BaseService from './BaseService';

export default class MarketDataService extends BaseService {
	public subProcesses: { [key: string]: ISubProcess } = {};

	public async startFetching(tool: string, option: IOption): Promise<void> {
		const assetIds = option.assets.length
			? option.assets
			: option.pair.split('|').filter(asset => asset !== '');

		if (!option.source) {
			util.logInfo('launching all sources for assets ' + assetIds.join(',') + ' in WS');
			for (const source in apis) {
				await util.sleep(1000);
				this.subProcesses[source] = {
					source: source,
					lastFailTimestamp: 0,
					failCount: 0,
					instance: undefined as any
				};
				this.launchSource(tool, source, assetIds, option);
			}
		} else {
			const api = apis[option.source];
			if (!api) throw new Error('invalid source');
			api.init();

			const sourceInstruments = api.getSourcePairs(assetIds);
			if (!sourceInstruments.length) {
				util.logInfo(`[${option.source}]: No sourceInstruments to fetch. Process.exit()`);
				return;
			}
			util.logInfo(
				`[${option.source}]:` +
					'start fetching for pairs ' +
					JSON.stringify(
						sourceInstruments.map(instrument => api.sourcePairMapping[instrument])
					)
			);
			apis[option.source].fetchTrades(sourceInstruments);
		}
	}

	public launchSource(tool: string, source: string, assets: string[], option: IOption) {
		util.logInfo(source);
		const cmd =
			`npm run ${tool} source=${source} assets=${assets.join(',')}${
				option.dynamo ? ' dynamo' : ''
			}${option.azure ? ' azure' : ''}${option.gcp ? ' gcp' : ''}${option.aws ? ' aws' : ''}${
				option.server ? ' server' : ''
			} ${option.forceREST ? 'rest' : ''}` +
			(osUtil.isWindows() ? '>>' : '&>') +
			` ${tool}.${source}.${assets.join('.')}.log`;

		util.logInfo(`[${source}]: ${cmd}`);

		const procInstance = child_process.exec(
			cmd,
			osUtil.isWindows() ? {} : { shell: '/bin/bash' }
		);

		this.subProcesses[source].instance = procInstance;
		this.subProcesses[source].lastFailTimestamp = util.getUTCNowTimestamp();

		if (!procInstance) {
			util.logError('Failed to launch ' + source);
			this.retry(tool, option, source, assets);
		} else {
			util.logInfo(`[${source}]: Launched ${tool}  ${assets.join(',')}`);
			procInstance.on('exit', code => {
				util.logError(`[${source}]: Exit with code ${code}`);
				if (code) this.retry(tool, option, source, assets);
			});
		}
	}

	public retry(tool: string, option: IOption, source: string, assets: string[]) {
		const now: number = util.getUTCNowTimestamp();

		if (now - this.subProcesses[source].lastFailTimestamp < 30000)
			this.subProcesses[source].failCount++;
		else this.subProcesses[source].failCount = 1;

		this.subProcesses[source].lastFailTimestamp = now;

		if (this.subProcesses[source].failCount < 3)
			global.setTimeout(() => this.launchSource(tool, source, assets, option), 5000);
		else util.logError('Retry Aborted ' + source);
	}

	public async cleanDb() {
		await dbUtil.init(this.tool, this.option, this.web3Wrapper);
		dbUtil.cleanDB();
		global.setInterval(() => dbUtil.cleanDB(), 60000 * 60 * 24);
		global.setInterval(() => dbUtil.insertHeartbeat(), 30000);
	}
}
