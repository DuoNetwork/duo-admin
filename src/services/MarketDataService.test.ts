// fix for @ledgerhq/hw-transport-u2f 4.28.0
import '@babel/polyfill';
import child_process from 'child_process';
import { kovan } from '../../../duo-contract-wrapper/src/contractAddresses';
import geminiApi from '../apis/geminiApi';
import dbUtil from '../utils/dbUtil';
import osUtil from '../utils/osUtil';
import util from '../utils/util';
import MarketDataService from './MarketDataService';
jest.mock('../../../duo-contract-wrapper/src/Web3Wrapper', () =>
	jest.fn(() => ({
		contractAddresses: kovan
	}))
);

const marketDataService = new MarketDataService('tool', {} as any);
test('retry after long enought time', () => {
	const launchMock = jest.fn();
	const launchOriginal = marketDataService.launchSource;
	marketDataService.launchSource = launchMock;
	child_process.exec = jest.fn() as any;
	util.getUTCNowTimestamp = jest.fn(() => 1234567890);
	global.setTimeout = jest.fn();
	marketDataService.subProcesses['source'] = {
		source: 'source',
		lastFailTimestamp: util.getUTCNowTimestamp() - (30000 + 1),
		failCount: 2,
		instance: undefined as any
	};
	marketDataService.retry(
		'tool',
		{
			forceREST: false,
			debug: false,
			live: false
		} as any,
		'source',
		['asset1', 'asset2', 'asset3']
	);
	expect(marketDataService.subProcesses['source']).toMatchSnapshot();
	expect((global.setTimeout as jest.Mock).mock.calls).toMatchSnapshot();
	(global.setTimeout as jest.Mock).mock.calls[0][0]();
	expect(launchMock.mock.calls).toMatchSnapshot();
	marketDataService.launchSource = launchOriginal;
});

test('retry within short time', () => {
	child_process.exec = jest.fn() as any;
	util.getUTCNowTimestamp = jest.fn(() => 1234567890);
	child_process.exec = jest.fn() as any;
	marketDataService.subProcesses['source'] = {
		source: 'source',
		lastFailTimestamp: util.getUTCNowTimestamp() - (30000 - 1),
		failCount: 2,
		instance: undefined as any
	};
	marketDataService.retry(
		'tool',
		{
			forceREST: false,
			debug: false,
			live: false
		} as any,
		'source',
		['asset1', 'asset2', 'asset3']
	);
	expect(marketDataService.subProcesses['source']).toMatchSnapshot();
});

test('launchSource fail windows', () => {
	osUtil.isWindows = jest.fn(() => true);
	child_process.exec = jest.fn() as any;
	util.getUTCNowTimestamp = jest.fn(() => 1234567890);
	marketDataService.retry = jest.fn();
	marketDataService.subProcesses['source'] = {
		source: 'source',
		lastFailTimestamp: 0,
		failCount: 0,
		instance: undefined as any
	};
	marketDataService.launchSource('tool', 'source', ['asset1', 'asset2', 'asset3'], {
		forceREST: false,
		debug: false,
		live: false,
		dynamo: true,
		azure: true
	} as any);
	expect(((child_process.exec as any) as jest.Mock).mock.calls).toMatchSnapshot();
	expect(marketDataService.subProcesses).toMatchSnapshot();
	expect((marketDataService.retry as jest.Mock<void>).mock.calls).toMatchSnapshot();
});

test('launchSource success windows', () => {
	osUtil.isWindows = jest.fn(() => true);
	const execOn = jest.fn();
	child_process.exec = jest.fn(() => {
		return {
			on: execOn
		};
	}) as any;
	util.getUTCNowTimestamp = jest.fn(() => 1234567890);
	marketDataService.retry = jest.fn();
	marketDataService.subProcesses['source'] = {
		source: 'source',
		lastFailTimestamp: 5,
		failCount: 0,
		instance: undefined as any
	};
	marketDataService.launchSource('tool', 'source', ['asset1', 'asset2', 'asset3'], {
		forceREST: false,
		debug: false,
		live: false,
		aws: true
	} as any);
	expect(((child_process.exec as any) as jest.Mock).mock.calls).toMatchSnapshot();
	expect(marketDataService.subProcesses).toMatchSnapshot();
	expect(execOn.mock.calls).toMatchSnapshot();
	execOn.mock.calls[0][1]();
	execOn.mock.calls[0][1](1);
	expect((marketDataService.retry as jest.Mock).mock.calls).toMatchSnapshot();
});

test('launchSource forceREST windows', () => {
	osUtil.isWindows = jest.fn(() => true);
	child_process.exec = jest.fn() as any;
	util.getUTCNowTimestamp = jest.fn(() => 1234567890);
	marketDataService.subProcesses['source'] = {
		source: 'source',
		lastFailTimestamp: 0,
		failCount: 0,
		instance: undefined as any
	};
	marketDataService.launchSource('tool', 'source', ['asset1', 'asset2', 'asset3'], {
		forceREST: true,
		debug: false,
		live: false,
		gcp: true,
		server: true
	} as any);
	expect(((child_process.exec as any) as jest.Mock).mock.calls).toMatchSnapshot();
	expect(marketDataService.subProcesses).toMatchSnapshot();
});

test('launchSource debug windows', () => {
	osUtil.isWindows = jest.fn(() => true);
	child_process.exec = jest.fn() as any;
	util.getUTCNowTimestamp = jest.fn(() => 1234567890);
	marketDataService.subProcesses['source'] = {
		source: 'source',
		lastFailTimestamp: 0,
		failCount: 0,
		instance: undefined as any
	};
	marketDataService.launchSource('tool', 'source', ['asset1', 'asset2', 'asset3'], {
		forceREST: false,
		debug: true,
		live: false
	} as any);
	expect(((child_process.exec as any) as jest.Mock).mock.calls).toMatchSnapshot();
	expect(marketDataService.subProcesses).toMatchSnapshot();
});

test('launchSource live windows', () => {
	osUtil.isWindows = jest.fn(() => true);
	child_process.exec = jest.fn() as any;
	util.getUTCNowTimestamp = jest.fn(() => 1234567890);
	marketDataService.subProcesses['source'] = {
		source: 'source',
		lastFailTimestamp: 0,
		failCount: 0,
		instance: undefined as any
	};
	marketDataService.launchSource('tool', 'source', ['asset1', 'asset2', 'asset3'], {
		forceREST: false,
		debug: false,
		live: true
	} as any);
	expect(((child_process.exec as any) as jest.Mock).mock.calls).toMatchSnapshot();
	expect(marketDataService.subProcesses).toMatchSnapshot();
});

test('launchSource fail not windows', () => {
	osUtil.isWindows = jest.fn(() => false);
	child_process.exec = jest.fn() as any;
	util.getUTCNowTimestamp = jest.fn(() => 1234567890);
	marketDataService.retry = jest.fn();
	marketDataService.subProcesses['source'] = {
		source: 'source',
		lastFailTimestamp: 0,
		failCount: 0,
		instance: undefined as any
	};
	marketDataService.launchSource('tool', 'source', ['asset1', 'asset2', 'asset3'], {
		forceREST: false,
		debug: false,
		live: false
	} as any);
	expect(((child_process.exec as any) as jest.Mock).mock.calls).toMatchSnapshot();
	expect(marketDataService.subProcesses).toMatchSnapshot();
	expect((marketDataService.retry as jest.Mock<void>).mock.calls).toMatchSnapshot();
});

test('launchSource success not windows', () => {
	osUtil.isWindows = jest.fn(() => false);
	child_process.exec = jest.fn(() => {
		return {
			on: jest.fn()
		};
	}) as any;
	util.getUTCNowTimestamp = jest.fn(() => 1234567890);
	marketDataService.retry = jest.fn();
	marketDataService.subProcesses['source'] = {
		source: 'source',
		lastFailTimestamp: 5,
		failCount: 0,
		instance: undefined as any
	};
	marketDataService.launchSource('tool', 'source', ['asset1', 'asset2', 'asset3'], {
		forceREST: false,
		debug: false,
		live: false
	} as any);
	expect(((child_process.exec as any) as jest.Mock).mock.calls).toMatchSnapshot();
	expect(marketDataService.subProcesses).toMatchSnapshot();
	expect(
		(marketDataService.subProcesses['source'].instance.on as jest.Mock<void>).mock.calls
	).toMatchSnapshot();
});

test('launchSource forceREST not windows', () => {
	osUtil.isWindows = jest.fn(() => false);
	child_process.exec = jest.fn() as any;
	util.getUTCNowTimestamp = jest.fn(() => 1234567890);
	marketDataService.subProcesses['source'] = {
		source: 'source',
		lastFailTimestamp: 0,
		failCount: 0,
		instance: undefined as any
	};
	marketDataService.launchSource('tool', 'source', ['asset1', 'asset2', 'asset3'], {
		forceREST: true,
		debug: false,
		live: false
	} as any);
	expect(((child_process.exec as any) as jest.Mock).mock.calls).toMatchSnapshot();
	expect(marketDataService.subProcesses).toMatchSnapshot();
});

test('launchSource debug not windows', () => {
	osUtil.isWindows = jest.fn(() => false);
	child_process.exec = jest.fn() as any;
	util.getUTCNowTimestamp = jest.fn(() => 1234567890);
	marketDataService.subProcesses['source'] = {
		source: 'source',
		lastFailTimestamp: 0,
		failCount: 0,
		instance: undefined as any
	};
	marketDataService.launchSource('tool', 'source', ['asset1', 'asset2', 'asset3'], {
		forceREST: false,
		debug: true,
		live: false
	} as any);
	expect(((child_process.exec as any) as jest.Mock).mock.calls).toMatchSnapshot();
	expect(marketDataService.subProcesses).toMatchSnapshot();
});

test('launchSource live not windows', () => {
	osUtil.isWindows = jest.fn(() => false);
	child_process.exec = jest.fn() as any;
	util.getUTCNowTimestamp = jest.fn(() => 1234567890);
	marketDataService.subProcesses['source'] = {
		source: 'source',
		lastFailTimestamp: 0,
		failCount: 0,
		instance: undefined as any
	};
	marketDataService.launchSource('tool', 'source', ['asset1', 'asset2', 'asset3'], {
		forceREST: false,
		debug: false,
		live: true
	} as any);
	expect(((child_process.exec as any) as jest.Mock).mock.calls).toMatchSnapshot();
	expect(marketDataService.subProcesses).toMatchSnapshot();
});

test('startFetching no source', async () => {
	util.sleep = jest.fn();
	marketDataService.launchSource = jest.fn();

	await marketDataService.startFetching('tool', {
		source: '',
		sources: [],
		exSources: [],
		assets: ['quote', 'base'],
		forceREST: false,
		debug: false
	} as any);

	await marketDataService.startFetching('tool', {
		source: '',
		sources: ['gemini', 'kraken'],
		exSources: [],
		assets: ['quote', 'base'],
		forceREST: false,
		debug: false
	} as any);

	await marketDataService.startFetching('tool', {
		source: '',
		sources: ['gemini', 'kraken'],
		exSources: [],
		pair: 'someQuote|someBase',
		assets: [],
		forceREST: false,
		debug: false
	} as any);

	expect((marketDataService.launchSource as jest.Mock).mock.calls).toMatchSnapshot();
});

test('startFetching source', async () => {
	util.sleep = jest.fn();
	geminiApi.getSourcePairs = jest.fn(() => ['quote|base']);
	geminiApi.fetchTrades = jest.fn();
	await marketDataService.startFetching('tool', {
		source: 'gemini',
		assets: ['quote', 'base'],
		forceREST: true,
		debug: false
	} as any);

	try {
		await marketDataService.startFetching('tool', {
			source: 'gemini1',
			sources: [],
			exSources: [],
			pair: 'someQuote|someBase',
			assets: [],
			forceREST: false,
			debug: false
		} as any);
	} catch (error) {
		expect(error).toMatchSnapshot();
	}

	expect((geminiApi.fetchTrades as jest.Mock).mock.calls).toMatchSnapshot();
});

test('startFetching source no sourcePairs', async () => {
	util.sleep = jest.fn();
	geminiApi.getSourcePairs = jest.fn(() => []);
	geminiApi.fetchTrades = jest.fn();
	await marketDataService.startFetching('tool', {
		source: 'gemini',
		assets: ['quote', 'base'],
		forceREST: true,
		debug: false
	} as any);

	expect(geminiApi.fetchTrades as jest.Mock).not.toBeCalled();
});

test('cleanDb', async () => {
	dbUtil.init = jest.fn();
	dbUtil.cleanDB = jest.fn();
	dbUtil.insertHeartbeat = jest.fn();
	global.setInterval = jest.fn();
	await marketDataService.cleanDb();
	expect(dbUtil.cleanDB as jest.Mock).toBeCalledTimes(1);
	for (const call of (global.setInterval as jest.Mock).mock.calls)
		expect(call[1]).toMatchSnapshot();

	(global.setInterval as jest.Mock).mock.calls[0][0]();
	expect(dbUtil.cleanDB as jest.Mock).toBeCalledTimes(2);
	(global.setInterval as jest.Mock).mock.calls[1][0]();
	expect(dbUtil.insertHeartbeat as jest.Mock).toBeCalledTimes(1);
});
