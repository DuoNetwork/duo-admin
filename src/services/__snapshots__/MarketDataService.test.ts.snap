// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cleanDb 1`] = `86400000`;

exports[`cleanDb 2`] = `30000`;

exports[`launchEvent  1`] = `
Array [
  Array [
    "npm run tool event=event  gcp&> tool.event.log",
    Object {
      "shell": "/bin/bash",
    },
  ],
]
`;

exports[`launchEvent  2`] = `
Object {
  "event": Object {
    "failCount": 0,
    "instance": undefined,
    "lastFailTimestamp": 1234567890,
    "source": "event",
  },
  "source": Object {
    "failCount": 0,
    "instance": undefined,
    "lastFailTimestamp": 1234567890,
    "source": "source",
  },
}
`;

exports[`launchEvent azure 1`] = `
Array [
  Array [
    "npm run tool event=event  azure&> tool.event.log",
    Object {
      "shell": "/bin/bash",
    },
  ],
]
`;

exports[`launchEvent azure 2`] = `Array []`;

exports[`launchEvent azure 3`] = `
Array [
  Array [
    "exit",
    [Function],
  ],
]
`;

exports[`launchEvent azure 4`] = `
Object {
  "event": Object {
    "failCount": 0,
    "instance": Object {
      "on": [MockFunction] {
        "calls": Array [
          Array [
            "exit",
            [Function],
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      },
    },
    "lastFailTimestamp": 1234567890,
    "source": "event",
  },
  "source": Object {
    "failCount": 0,
    "instance": undefined,
    "lastFailTimestamp": 1234567890,
    "source": "source",
  },
}
`;

exports[`launchEvent azure 5`] = `
Array [
  Array [
    "tool",
    "event",
    Object {
      "azure": true,
      "debug": false,
      "event": "event",
      "forceREST": false,
      "live": true,
    },
  ],
]
`;

exports[`launchEvent with existingInstance, no windows 1`] = `
Array [
  Array [
    "Others",
    [Function],
  ],
]
`;

exports[`launchEvent with existingInstance, no windows 2`] = `
Array [
  Array [
    "tool",
    "Others",
    Object {
      "aws": true,
      "debug": false,
      "event": "Others",
      "forceREST": false,
      "live": true,
      "server": true,
    },
  ],
]
`;

exports[`launchEvent with existingInstance, no windows 3`] = `
Object {
  "Others": Object {
    "failCount": 0,
    "instance": null,
    "lastFailTimestamp": 1234567890,
    "source": "Others",
  },
  "event": Object {
    "failCount": 0,
    "instance": Object {
      "on": [MockFunction] {
        "calls": Array [
          Array [
            "exit",
            [Function],
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      },
    },
    "lastFailTimestamp": 1234567890,
    "source": "event",
  },
  "source": Object {
    "failCount": 0,
    "instance": undefined,
    "lastFailTimestamp": 1234567890,
    "source": "source",
  },
}
`;

exports[`launchSource debug not windows 1`] = `
Array [
  Array [
    "npm run tool source=source assets=asset1,asset2,asset3 &> tool.source.asset1.asset2.asset3.log",
    Object {
      "shell": "/bin/bash",
    },
  ],
]
`;

exports[`launchSource debug not windows 2`] = `
Object {
  "source": Object {
    "failCount": 0,
    "instance": undefined,
    "lastFailTimestamp": 1234567890,
    "source": "source",
  },
}
`;

exports[`launchSource debug windows 1`] = `
Array [
  Array [
    "npm run tool source=source assets=asset1,asset2,asset3 >> tool.source.asset1.asset2.asset3.log",
    Object {},
  ],
]
`;

exports[`launchSource debug windows 2`] = `
Object {
  "source": Object {
    "failCount": 0,
    "instance": undefined,
    "lastFailTimestamp": 1234567890,
    "source": "source",
  },
}
`;

exports[`launchSource fail not windows 1`] = `
Array [
  Array [
    "npm run tool source=source assets=asset1,asset2,asset3 &> tool.source.asset1.asset2.asset3.log",
    Object {
      "shell": "/bin/bash",
    },
  ],
]
`;

exports[`launchSource fail not windows 2`] = `
Object {
  "source": Object {
    "failCount": 0,
    "instance": undefined,
    "lastFailTimestamp": 1234567890,
    "source": "source",
  },
}
`;

exports[`launchSource fail not windows 3`] = `
Array [
  Array [
    "source",
    [Function],
  ],
]
`;

exports[`launchSource fail windows 1`] = `
Array [
  Array [
    "npm run tool source=source assets=asset1,asset2,asset3 dynamo azure >> tool.source.asset1.asset2.asset3.log",
    Object {},
  ],
]
`;

exports[`launchSource fail windows 2`] = `
Object {
  "source": Object {
    "failCount": 0,
    "instance": false,
    "lastFailTimestamp": 1234567890,
    "source": "source",
  },
}
`;

exports[`launchSource fail windows 3`] = `
Array [
  Array [
    "source",
    [Function],
  ],
]
`;

exports[`launchSource fail windows 4`] = `
Array [
  Array [
    "tool",
    "source",
    Array [
      "asset1",
      "asset2",
      "asset3",
    ],
    Object {
      "azure": true,
      "debug": false,
      "dynamo": true,
      "forceREST": false,
      "live": false,
    },
  ],
]
`;

exports[`launchSource forceREST not windows 1`] = `
Array [
  Array [
    "npm run tool source=source assets=asset1,asset2,asset3 rest&> tool.source.asset1.asset2.asset3.log",
    Object {
      "shell": "/bin/bash",
    },
  ],
]
`;

exports[`launchSource forceREST not windows 2`] = `
Object {
  "source": Object {
    "failCount": 0,
    "instance": undefined,
    "lastFailTimestamp": 1234567890,
    "source": "source",
  },
}
`;

exports[`launchSource forceREST windows 1`] = `
Array [
  Array [
    "npm run tool source=source assets=asset1,asset2,asset3 gcp server rest>> tool.source.asset1.asset2.asset3.log",
    Object {},
  ],
]
`;

exports[`launchSource forceREST windows 2`] = `
Object {
  "source": Object {
    "failCount": 0,
    "instance": undefined,
    "lastFailTimestamp": 1234567890,
    "source": "source",
  },
}
`;

exports[`launchSource live not windows 1`] = `
Array [
  Array [
    "npm run tool source=source assets=asset1,asset2,asset3 &> tool.source.asset1.asset2.asset3.log",
    Object {
      "shell": "/bin/bash",
    },
  ],
]
`;

exports[`launchSource live not windows 2`] = `
Object {
  "source": Object {
    "failCount": 0,
    "instance": undefined,
    "lastFailTimestamp": 1234567890,
    "source": "source",
  },
}
`;

exports[`launchSource live windows 1`] = `
Array [
  Array [
    "npm run tool source=source assets=asset1,asset2,asset3 >> tool.source.asset1.asset2.asset3.log",
    Object {},
  ],
]
`;

exports[`launchSource live windows 2`] = `
Object {
  "source": Object {
    "failCount": 0,
    "instance": undefined,
    "lastFailTimestamp": 1234567890,
    "source": "source",
  },
}
`;

exports[`launchSource success not windows 1`] = `
Array [
  Array [
    "npm run tool source=source assets=asset1,asset2,asset3 &> tool.source.asset1.asset2.asset3.log",
    Object {
      "shell": "/bin/bash",
    },
  ],
]
`;

exports[`launchSource success not windows 2`] = `
Object {
  "source": Object {
    "failCount": 0,
    "instance": Object {
      "on": [MockFunction] {
        "calls": Array [
          Array [
            "exit",
            [Function],
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      },
    },
    "lastFailTimestamp": 1234567890,
    "source": "source",
  },
}
`;

exports[`launchSource success not windows 3`] = `
Array [
  Array [
    "exit",
    [Function],
  ],
]
`;

exports[`launchSource success windows 1`] = `
Array [
  Array [
    "npm run tool source=source assets=asset1,asset2,asset3 aws >> tool.source.asset1.asset2.asset3.log",
    Object {},
  ],
]
`;

exports[`launchSource success windows 2`] = `
Object {
  "source": Object {
    "failCount": 0,
    "instance": Object {
      "on": [MockFunction] {
        "calls": Array [
          Array [
            "exit",
            [Function],
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      },
    },
    "lastFailTimestamp": 1234567890,
    "source": "source",
  },
}
`;

exports[`launchSource success windows 3`] = `
Array [
  Array [
    "exit",
    [Function],
  ],
]
`;

exports[`launchSource success windows 4`] = `
Array [
  Array [
    "source",
    [Function],
  ],
]
`;

exports[`launchSource success windows 5`] = `
Array [
  Array [
    "tool",
    "source",
    Array [
      "asset1",
      "asset2",
      "asset3",
    ],
    Object {
      "aws": true,
      "debug": false,
      "forceREST": false,
      "live": false,
    },
  ],
]
`;

exports[`retry after long enought time 1`] = `
Object {
  "failCount": 1,
  "instance": undefined,
  "lastFailTimestamp": 1234567890,
  "source": "source",
}
`;

exports[`retry after long enought time 2`] = `
Array [
  Array [
    [Function],
    5000,
  ],
]
`;

exports[`retry within short time 1`] = `
Object {
  "failCount": 3,
  "instance": undefined,
  "lastFailTimestamp": 1234567890,
  "source": "source",
}
`;

exports[`startAggregate 1`] = `
Array [
  Array [
    Object {
      "period": Object {
        "N": "10",
      },
    },
  ],
  Array [
    Object {
      "period": Object {
        "N": "10",
      },
    },
  ],
]
`;

exports[`startAggregate 2`] = `
Array [
  Array [
    10,
  ],
  Array [
    10,
  ],
]
`;

exports[`startFetching no source 1`] = `
Array [
  Array [
    "tool",
    "gemini",
    Array [
      "quote",
      "base",
    ],
    Object {
      "assets": Array [
        "quote",
        "base",
      ],
      "debug": false,
      "exSources": Array [],
      "forceREST": false,
      "source": "",
      "sources": Array [],
    },
  ],
  Array [
    "tool",
    "kraken",
    Array [
      "quote",
      "base",
    ],
    Object {
      "assets": Array [
        "quote",
        "base",
      ],
      "debug": false,
      "exSources": Array [],
      "forceREST": false,
      "source": "",
      "sources": Array [],
    },
  ],
  Array [
    "tool",
    "gdax",
    Array [
      "quote",
      "base",
    ],
    Object {
      "assets": Array [
        "quote",
        "base",
      ],
      "debug": false,
      "exSources": Array [],
      "forceREST": false,
      "source": "",
      "sources": Array [],
    },
  ],
  Array [
    "tool",
    "bitstamp",
    Array [
      "quote",
      "base",
    ],
    Object {
      "assets": Array [
        "quote",
        "base",
      ],
      "debug": false,
      "exSources": Array [],
      "forceREST": false,
      "source": "",
      "sources": Array [],
    },
  ],
  Array [
    "tool",
    "gemini",
    Array [
      "quote",
      "base",
    ],
    Object {
      "assets": Array [
        "quote",
        "base",
      ],
      "debug": false,
      "exSources": Array [],
      "forceREST": false,
      "source": "",
      "sources": Array [
        "gemini",
        "kraken",
      ],
    },
  ],
  Array [
    "tool",
    "kraken",
    Array [
      "quote",
      "base",
    ],
    Object {
      "assets": Array [
        "quote",
        "base",
      ],
      "debug": false,
      "exSources": Array [],
      "forceREST": false,
      "source": "",
      "sources": Array [
        "gemini",
        "kraken",
      ],
    },
  ],
  Array [
    "tool",
    "gdax",
    Array [
      "quote",
      "base",
    ],
    Object {
      "assets": Array [
        "quote",
        "base",
      ],
      "debug": false,
      "exSources": Array [],
      "forceREST": false,
      "source": "",
      "sources": Array [
        "gemini",
        "kraken",
      ],
    },
  ],
  Array [
    "tool",
    "bitstamp",
    Array [
      "quote",
      "base",
    ],
    Object {
      "assets": Array [
        "quote",
        "base",
      ],
      "debug": false,
      "exSources": Array [],
      "forceREST": false,
      "source": "",
      "sources": Array [
        "gemini",
        "kraken",
      ],
    },
  ],
  Array [
    "tool",
    "gemini",
    Array [
      "someQuote",
      "someBase",
    ],
    Object {
      "assets": Array [],
      "debug": false,
      "exSources": Array [],
      "forceREST": false,
      "pair": "someQuote|someBase",
      "source": "",
      "sources": Array [
        "gemini",
        "kraken",
      ],
    },
  ],
  Array [
    "tool",
    "kraken",
    Array [
      "someQuote",
      "someBase",
    ],
    Object {
      "assets": Array [],
      "debug": false,
      "exSources": Array [],
      "forceREST": false,
      "pair": "someQuote|someBase",
      "source": "",
      "sources": Array [
        "gemini",
        "kraken",
      ],
    },
  ],
  Array [
    "tool",
    "gdax",
    Array [
      "someQuote",
      "someBase",
    ],
    Object {
      "assets": Array [],
      "debug": false,
      "exSources": Array [],
      "forceREST": false,
      "pair": "someQuote|someBase",
      "source": "",
      "sources": Array [
        "gemini",
        "kraken",
      ],
    },
  ],
  Array [
    "tool",
    "bitstamp",
    Array [
      "someQuote",
      "someBase",
    ],
    Object {
      "assets": Array [],
      "debug": false,
      "exSources": Array [],
      "forceREST": false,
      "pair": "someQuote|someBase",
      "source": "",
      "sources": Array [
        "gemini",
        "kraken",
      ],
    },
  ],
]
`;

exports[`startFetching source 1`] = `[Error: invalid source]`;

exports[`startFetching source 2`] = `
Array [
  Array [
    Array [
      "quote|base",
    ],
  ],
]
`;

exports[`startFetchingEvent no event 1`] = `
Array [
  Array [
    "tool",
    "event1",
    Object {
      "debug": false,
      "event": "event2",
      "events": Array [
        "event1",
        "event2",
      ],
    },
  ],
  Array [
    "tool",
    "event2",
    Object {
      "debug": false,
      "event": "event2",
      "events": Array [
        "event1",
        "event2",
      ],
    },
  ],
]
`;

exports[`startFetchingEvent no event 2`] = `
Object {
  "Others": Object {
    "failCount": 0,
    "instance": null,
    "lastFailTimestamp": 1234567890,
    "source": "Others",
  },
  "bitstamp": Object {
    "failCount": 0,
    "instance": undefined,
    "lastFailTimestamp": 0,
    "source": "bitstamp",
  },
  "event": Object {
    "failCount": 0,
    "instance": Object {
      "on": [MockFunction] {
        "calls": Array [
          Array [
            "exit",
            [Function],
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      },
    },
    "lastFailTimestamp": 1234567890,
    "source": "event",
  },
  "event1": Object {
    "failCount": 0,
    "instance": undefined,
    "lastFailTimestamp": 0,
    "source": "event1",
  },
  "event2": Object {
    "failCount": 0,
    "instance": undefined,
    "lastFailTimestamp": 0,
    "source": "event2",
  },
  "gdax": Object {
    "failCount": 0,
    "instance": undefined,
    "lastFailTimestamp": 0,
    "source": "gdax",
  },
  "gemini": Object {
    "failCount": 0,
    "instance": undefined,
    "lastFailTimestamp": 0,
    "source": "gemini",
  },
  "kraken": Object {
    "failCount": 0,
    "instance": undefined,
    "lastFailTimestamp": 0,
    "source": "kraken",
  },
  "source": Object {
    "failCount": 0,
    "instance": undefined,
    "lastFailTimestamp": 1234567890,
    "source": "source",
  },
}
`;

exports[`startFetchingEvent with event 1`] = `
Array [
  Array [
    "tool",
    Object {
      "debug": false,
      "event": "event",
    },
  ],
  Array [
    "tool",
    Object {
      "debug": false,
      "event": "StartReset",
    },
  ],
]
`;
